mods/default/item_entity.lua

```diff
@@ -0,0 +1,64 @@
+-- mods/default/item_entity.lua
+
+local builtin_item = minetest.registered_entities["__builtin:item"]
+
+local item = {
+	set_item = function(self, itemstring)
+		builtin_item.set_item(self, itemstring)
+
+		local stack = ItemStack(itemstring)
+		local itemdef = minetest.registered_items[stack:get_name()]
+		if itemdef then
+			self.flammable = itemdef.groups.flammable
+		end
+	end,
+
+	burn_up = function(self)
+		-- disappear in a smoke puff
+		self.object:remove()
+		local p = self.object:getpos()
+		minetest.add_particlespawner({
+			amount = 3,
+			time = 0.1,
+			minpos = {x=p.x, y=p.y, z=p.z},
+			maxpos = {x=p.x, y=p.y+0.2, z=p.z},
+			minacc = {x=-0.5,y=5,z=-0.5},
+			maxacc = {x=0.5,y=5,z=0.5},
+			minexptime = 0.1,
+			minsize = 2,
+			maxsize = 4,
+			texture = "smoke_puff.png"
+		})
+	end,
+
+	on_step = function(self, dtime)
+		builtin_item.on_step(self, dtime)
+
+		if self.flammable then
+			-- flammable, check for igniters
+			self.ignite_timer = (self.ignite_timer or 0) + dtime
+			if self.ignite_timer > 10 then
+				self.ignite_timer = 0
+
+				local node = minetest.get_node(self.object:getpos())
+
+				-- Immediatelly burn up flammable items in lava
+				if minetest.get_item_group(node.name, "lava") > 0 then
+					self:burn_up()
+				else
+					--  otherwise there'll be a chance based on its igniter value
+					local burn_chance = self.flammable
+						* minetest.get_item_group(node.name, "igniter")
+					if burn_chance > 0 and math.random(0, burn_chance) ~= 0 then
+						self:burn_up()
+					end
+				end
+			end
+		end
+	end,
+}
+
+-- set defined item as new __builtin:item, with the old one as fallback table
+setmetatable(item, builtin_item)
```

Comment:
All registered entities have the `__index` metamethod set up, [it's part of `core.register_entity`](https://github.com/minetest/minetest/blob/stable-0.4/builtin/game/register.lua#L98), which is used before in builtin for the builtin_item.